lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -1,10 +1,164 @@
+    ~r{
+      # TODO: revise the erlang rules for function names
+      [0-9a-zA-Z_!\\?]+
+    }x
+  end
+
+  def re(:fa, language) when language in [:elixir, :erlang] do
+    ~r{
+      (#{re_source(:f, language)})            # function name
+      /\d+                                    # slash + arity
+    }x
+  end
+
+  def re(:mfa, :elixir) do
+    ~r{
+      (#{re_source(:prefix)})?                # optional callback/type identifier or ":"
+      (
+        (#{re_source(:m)}\.)
+        #{re_source(:fa)}
+      )
+    }x
+  end
+
+  def re(:mfa, :erlang) do
+    ~r{
+      #{re_source(:m, :erlang)}               # module name
+      \.                                      # "."
+      #{re_source(:fa, :erlang)}              # function name
+    }x
+  end
+
+  def re(:local, :elixir) do
+    ~r{
+      (#{re_source(:prefix)})?               # optional callback or type identifier
+      #{re_source(:fa)}                      # function name + arity
+    }x
+  end
+
+  def re(:modules, :elixir) do
+    ~r{
+      #{re_source(:m)}
+    }x
+  end
+
+  def re(:modules, :erlang) do
+    ~r{
+      #{re_source(:m, :erlang)}
+    }x
+  end
+
+  def re(:functions, :elixir) do
+    ~r{
+      (#{re_source(:local)}) | (#{re_source(:mfa)})
+    }x
+  end
+
+  def re(:functions, :erlang) do
+    ~r{
+      #{re_source(:mfa, :erlang)}
+    }x
+  end
+
+  def re({:normal_link, function_re_source}, :markdown) do
+    ~r{
+      (?<!\[)(?<!\()                            # it shouldn't be preceded by "[", "("
+      (?<!``)
+      `\s*                                      # leading backtick
+      (#{function_re_source})                   # CAPTURE 1
+      \s*`                                      # trailing backtick
+      (?!`)
+      (?!\])(?!\))                              # it shouldn't be followed by "]", ")"
+    }x
+  end
+
+  def re({:custom_link, function_re_source}, :markdown) do
+    re({:custom_link, "(.*?)", function_re_source}, :markdown)
+  end
+
+  def re({:custom_link, text_re_source, function_re_source}, :markdown) do
+    ~r{
+      \[#{text_re_source}\]                     # CAPTURE 1
+      \(`(#{function_re_source})`\)             # CAPTURE 2
+    }x
+  end
+
+  def regex_link_type(language, kind, link_type) do
+    group =
+      case kind do
+        :function -> :functions
+        :module -> :modules
+      end
+
+    case link_type do
+      :normal ->
+        re({:normal_link, re_source(group, language)}, :markdown)
+
+      :custom ->
+        re({:custom_link, re_source(group, language)}, :markdown)
+    end
+  end
+end
+
 defmodule ExDoc.Formatter.HTML.Autolink do
   @moduledoc """
   Conveniences for autolinking.
   """
-
+  import ExDoc.SyntaxRule
   import ExDoc.Formatter.HTML.Templates, only: [h: 1, enc_h: 1]
 
+  @type language :: ExDoc.SyntaxRule.language()
+  @type kind :: :function | :module
+  @type link_type :: :normal | :custom
+
+  random_string = 1..10_000_000 |> Enum.random() |> Integer.to_charlist(36)
```

Comment:
Let's pick a random string. Having it being a different value every time we compile the code will make it hard to spot bugs.